torch/quantization/convert_modules.py

```diff
@@ -0,0 +1,222 @@
+            module_prefix = prefix + '.' + name if prefix else name
+            add_observer(child, qconfig_dict, module.qconfig, module_prefix)
+
+def add_quant_dequant_call(base_fn):
+    r"""Insert calls to quant and dequant around base_fn calls
+
+    Given a `base_fn` function, insert a call to `quant` before the function
+    call which is supposed to quantize the input and a call to `dequant` after
+    the function call which will dequantize the output of the function.
+
+    Args:
+        base_fn: base function we want to call
+
+    Returns:
+        new_fn: a modified function which calls quant -> base_bn -> dequant
+    """
+    def new_fn(x):
+        x = base_fn.__self__.quant(x)
+        x = base_fn(x)
+        x = base_fn.__self__.dequant(x)
+        return x
+    return new_fn
+
+class QuantWrapper(nn.Module):
+    def __init__(self, module, qconfig):
+        super(QuantWrapper, self).__init__()
+        self.quant = QuantStub(qconfig)
+        self.dequant = DeQuantStub()
+        self.module = module
+
+    def forward(self, X):
+        X = self.quant(X)
+        self.module.forward(X)
+        X = self.dequant(X)
+        return X
+
+def get_config_key(name, qconfig_dict):
+    if name in qconfig_dict:
+        return name
+    elif name:
+        parent = '.'.join(name.split('.')[:-1])
+        return get_config_key(parent, qconfig_dict)
+    else:
+        return '' if '' in qconfig_dict else None
+
+def get_module(mod, name):
+    if name == '':
+        return mod
+    splits = name.split('.')
+    for split in splits:
+        mod = mod._modules[split]
+    return mod
+
+def add_quant_dequant_module(module, qconfig_dict, name):
+    assert name != ''
+    qconfig = qconfig_dict[name]
+    splits = name.split('.')
+    print('splits:', splits)
+    for split in splits:
+        parent_module = module
+        module = module._modules[split]
+    assert parent_module is not module
+    parent_module._modules[splits[-1]] = QuantWrapper(module, qconfig)
+
+def add_quant_dequant(module, qconfig_dict):
+    r"""Add QuantStub and DeQuantStub module and Wrap the module in QuantWrapper
+    according to qconfig_dict
+    """
+    mod_key_set = set()
+    for name, _ in module.named_modules():
+        print('name', name)
+        dict_key = get_config_key(name, qconfig_dict)
+        if dict_key is not None:
+            mod_key_set.add(dict_key)
+
+    # We need to update the modules from inner most
+    # module to outer modules(in decreasing order of number of '.')
+    print('mod_key_set', mod_key_set)
+    mod_key_heap = []
+    for name in mod_key_set:
+        heapq.heappush(mod_key_heap, (-name.count('.'), name))
+    while mod_key_heap:
+        _, name = heapq.heappop(mod_key_heap)
+        if name == '':
+            module = QuantWrapper(module, qconfig_dict[name])
+        else:
+            add_quant_dequant_module(module, qconfig_dict, name)
+    return module
+
+def prepare(module, qconfig_dict):
```

Comment:
should it be called `observe_and_quantize`?

Reply:
It is not quantizing here, only preparing for quantization. Also, if we want the same names for QAT, better to leave it at prepare: In the QAT case, prepare would insert fake-quants.